src/persist/src/indexed/mod.rs

```diff
@@ -1258,17 +1270,46 @@ impl AppliedState {
 
         Ok(())
     }
+}
 
-    /// Atomically moves all writes in unsealed not in advance of the trace's
-    /// seal frontier into the trace and does any necessary resulting eviction
-    /// work to remove unnecessary batches.
-    fn drain_unsealed<B: Blob>(&mut self, blob: &mut BlobCache<B>) -> Result<(), Error> {
-        for arrangement in self.arrangements.values_mut() {
-            arrangement.unsealed_drain(blob)?;
+impl<L: Log, B: Blob> Indexed<L, B> {
+    fn drain_unsealed(&mut self) -> Result<(), Error> {
+        self.validate_pending_empty()?;
+
+        // Do all the fallible work up front.
+        let mut responses = Vec::new();
+        for (id, arrangement) in self.state.arrangements.iter() {
+            if let Some(req) = arrangement.unsealed_next_drain_req()? {
+                let res = Arrangement::drain_unsealed_blocking(&self.blob, req)?;
+                responses.push((*id, res));
+            }
         }
+
+        self.apply_batched_cmd(|state, pending| {
+            for (id, res) in responses {
+                let arrangement = match state.arrangements.get_mut(&id) {
+                    Some(x) => x,
+                    None => continue,
+                };
+                if let Some(batch) = arrangement.unsealed_handle_drain_response(res) {
+                    pending.add_records_for_listeners(id, batch.updates.clone());
+                    // WIP can ListenEvent::Sealed be an Antichain?
+                    for element in batch.desc.upper().elements() {
+                        pending.add_seal_for_listener(id, *element);
+                    }
+                }
+            }
+        });
+
+        // Immediately drain pending so we don't have to reason about these
+        // floating around at the same time as pending writes and seals.
+        self.drain_pending()?;
```

Comment:
if we're draining pending right after applying an unbatched command then it kind of feels like a batched command

Reply:
ok yeah - it seems like we drain pending immediately before and immediately after this batched cmd, (looking at how drain_unsealed is used by step)

Reply:
did you mean to swap batched and unbatched command above? my guess is yes, so responding assuming that

I also initially thought that this should be an unbatched command, but unbatched commands don't have access to pending, and shoudn't. we don't really have a way to indicate to unbatched command, "here's a thing you should do if the meta write is a success". we could add something like that but...

- 1) i think it's uncontroversial that at some point we want to move unsealed_drain out of the hot loop
- 2) the existing background example (trace_compact) uses batched_command to process the response
- 3) I _think_ the `self.drain_pending()` call you left this comment on isn't necessary for correctness, but it seemed easier to me to reason about having it here and investigating removing it later

> I'm a little bit shaky on why we need to do the request / response thing for drain unsealed, and what sending data to the listeners as a batched cmd provides (as opposed to keeping drain unsealed a single function and then sending data to listeners at the end) but I haven't thought about this very deeply.

we don't _need_ to the req/res thing, the first version of this PR I wrote plumbed blob to the batched command closure and manually handled blob errors by snapshotting state at the beginning and rolling it back. but that's super gross (probably wouldn't merge it like that, instead would add a third variant to the batced/unbatched command structure) and given that everything works out super nicely with the req/res structure and we likely want to eventually do that anyway, that's how I ended up here

Reply:
> did you mean to swap batched and unbatched command above? my guess is yes

Indeed I did!